README.md
# strace
基于strace-4.14修改，增加以下功能


## 定时统计系统调用数 ##
如果使用-c参数统计系统调用不足之处在于，统计是从strace启动到停止整个进程生命周期的数据，不够精细化，有时需要像vmstat/mpstat之类的定时输出结果。
因此，新增-W n参数，用于设置定时统计输出的间隔时间，单位秒。

```
strace -W 1 -p 12345
------------------------------21:09:25-----------------------------
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 57.34    0.008547          21       398           poll
 19.38    0.002889           6       464       394 recvmsg
 14.99    0.002235          20       112         8 futex
  4.70    0.000701          10        72         1 read
  3.35    0.000499           8        65           writev
  0.16    0.000024           6         4           write
  0.08    0.000012           4         3           clone
------ ----------- ----------- --------- --------- ----------------
100.00    0.014907                  1118       403 total
------------------------------21:09:26-----------------------------
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 45.17    0.013233          39       341        14 futex
 21.28    0.006235           7       878           poll
 20.65    0.006049           5      1246       883 recvmsg
  7.42    0.002173           6       341           writev
  2.58    0.000755          18        41           read
  2.42    0.000708          15        46           write
  0.16    0.000048          24         2           open
  0.13    0.000038          38         1           clone
  0.09    0.000027           9         3           mmap
  0.05    0.000016          16         1           mprotect
  0.04    0.000012           6         2           close
------ ----------- ----------- --------- --------- ----------------
100.00    0.029294                  2902       897 total
------------------------------21:09:27-----------------------------
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 77.40    0.023020          30       777        99 futex
 13.09    0.003894          18       211           write
  3.32    0.000986           4       263           poll
  2.92    0.000869           3       261       248 recvmsg
  1.02    0.000303           7        41         1 read
  0.95    0.000282          35         8           mmap
  0.61    0.000180          36         5           open
  0.46    0.000136           6        21           writev
  0.11    0.000032          32         1           clone
  0.07    0.000021           4         5           close
  0.03    0.000008           8         1           mprotect
  0.02    0.000006           6         1           fstat
  0.02    0.000006           6         1           fcntl
  0.00    0.000000           0         1           munmap
  0.00    0.000000           0         5           getsockname
------ ----------- ----------- --------- --------- ----------------
100.00    0.029743                  1602       348 total
```


debian/changelog
strace (4.14-1) experimental; urgency=low

  * strace 4.14 snapshot.

 -- Strace <strace-devel@lists.sourceforge.net>  Tue, 04 Oct 2016 19:13:22 +0000

strace (4.13-1) unstable; urgency=medium

  * New upstream version.

 -- Dmitry V. Levin <ldv@altlinux.org>  Tue, 26 Jul 2016 16:17:18 +0000

strace (4.12-3) unstable; urgency=medium

  * Fix up 32/64-bit test output in strace -S test to fix i386 build
  * Run make check with VERBOSE=1 so we get useful information in build logs

 -- Steve McIntyre <93sam@debian.org>  Sat, 04 Jun 2016 02:37:44 +0100

strace (4.12-2) unstable; urgency=medium

  * Remove libunwind dependency. Closes: #826223

 -- Steve McIntyre <93sam@debian.org>  Fri, 03 Jun 2016 13:43:44 +0100

strace (4.12-1) unstable; urgency=low

  [ Dmitry V. Levin ]
  * New upstream version.
    + implemented simultaneous use of -p option and tracing of a command,
      closes: #549942
    + fixed decoding of syscalls unknown to the kernel on s390/s390x,
      closes: #485979

 -- Dmitry V. Levin <ldv@altlinux.org>  Tue, 31 May 2016 11:12:13 +0000

strace (4.11-1) unstable; urgency=low

  [ Dmitry V. Levin ]
  * New upstream version.
    + fixed decoding of path argument of umount2 syscall, closes: #785050

  [ Steve McIntyre ]
  * Drop patches against old upstream - all upstream now.
  * Add simple upstream patch to fix printing of 32-bit times syscall
    return value on 64-bit architectures.

 -- Steve McIntyre <93sam@debian.org>  Fri, 08 Jan 2015 00:52:05 +0000

strace (4.10-3) unstable; urgency=medium

  * Really fix mips* builds - fix sys_syscall() handling.
    Closes: #781753
  * Clean up strace64.substvars at clean time
  * Remove ancient changelog entries that annoy lintian
  * Update Standards-Version to 3.9.6 (no changes needed)

 -- Steve McIntyre <93sam@debian.org>  Wed, 08 Jul 2015 23:44:03 +0100

strace (4.10-2) unstable; urgency=medium

  * Fix building on arm64 using patches from upstream. Closes: #790770
  * Fix various test suite failures using patches from upstream.
    Closes: #781040
  * Fix mips o32 build. Thanks to Arturo Borrero Gonzalez
    <arturo.borrero.glez@gmail.com> for tracking this.
  * Add build-dependency on gawk for the test suite. Closes: #780992

 -- Steve McIntyre <93sam@debian.org>  Thu, 02 Jul 2015 02:03:25 +0100

strace (4.10-1) unstable; urgency=low

  [ Dmitry V. Levin ]
  * New upstream version.
    + enhanced ioctl decoding,  closes: #692913
    + enabled mips64el build, closes: #773375
    + updated signal and errno numbers, closes: #768642
    + updated generic and added arch-specific PTRACE_* constants,
      closes: #734554

 -- Steve McIntyre <93sam@debian.org>  Sat, 21 Mar 2015 22:18:25 +0000

strace (4.9-2) unstable; urgency=medium

  * Turn on support for ppc64el, code is already in upstream.
    Closes: #744121.

 -- Steve McIntyre <93sam@debian.org>  Wed, 20 Aug 2014 18:00:20 +0100

strace (4.9-1) unstable; urgency=medium

  [ Dmitry V. Levin ]
  * New upstream version.
    + added -w option to produce stats on syscall latency, closes: #457497
    + applied miscellaneous corrections to the manpage, closes: #725987
    + fixed a typo in package description, closes: #697625
    + enabled or1k build, closes: #742235
    + enabled maximum security hardening, thanks to Markus <waldeck@gmx.de>

  [ Steve McIntyre ]
  * New maintainer. Thanks to Frederik for all his prior work!

 -- Steve McIntyre <93sam@debian.org>  Wed, 20 Aug 2014 15:11:36 +0100

strace (4.8-1.1) unstable; urgency=low

  * NMU
  * Uploading 4.8 to unstable to fix long-standing bugs that have been
    fixed for ages in experimental only.
  * enable arm64 build, Closes: #749956
  * enable x32 build, Closes: #727018
  * lintian-inspired fixes:
    + add build-arch and build-indep targets in debian/rules
    + update standards-version to 3.9.5
    + remove ancient changelog entries causing syntax checker errors
    + add hardening using buildflags.mk
    + trivial changes to Description: fields

 -- Steve McIntyre <93sam@debian.org>  Thu, 10 Jul 2014 14:22:18 +0000

strace (4.8-1) experimental; urgency=low

  * New upstream version.
    + Added aarch64 support, closes: #693074
    + Fixed kernel release string parsing, closes: #702309

 -- Dmitry V. Levin <ldv@altlinux.org>  Mon, 03 Jun 2013 15:16:17 +0000

strace (4.7-1) experimental; urgency=low

  [ Dmitry V. Levin ]
  * New upstream version.
    + Fixed sockaddr_un.sun_path name in decoded output, closes: #554946
    + Avoid potential core file clobbering on exit, closes: #656398
    + Fixed a typo in documentation, closes: #653309

 -- Frederik Schüler <fs@debian.org>  Tue, 12 Feb 2013 10:35:22 +0100

strace (4.6-1) unstable; urgency=low

  * New upstream version.
    + Added HDIO_* ioctl names, closes: #450953
    + Fixed stat64 decoding on mips, closes: #599028
    + Fixed misleading italics in the manual page, closes: #589323

 -- Dmitry V. Levin <ldv@altlinux.org>  Mon, 14 Mar 2011 15:16:17 +0000

strace (4.5.20-2.3) unstable; urgency=low

  * Non-maintainer upload.
  * Add sh4 to arch list.
  * Add powerpcspe to arch list (Closes: #579842).

 -- Nobuhiro Iwamatsu <iwamatsu@debian.org>  Tue, 22 Mar 2011 09:03:49 +0900

strace (4.5.20-2.2) unstable; urgency=low

  * Non-maintainer upload.
  * Add sh4 to arch list (Closes: #537105).

 -- Nobuhiro Iwamatsu <iwamatsu@debian.org>  Wed, 09 Mar 2011 09:22:01 +0900

strace (4.5.20-2.1) unstable; urgency=low

  * NMU, added armhf support. (Closes: #596960)

 -- Konstantinos Margaritis <markos@debian.org>  Thu, 24 Feb 2011 14:47:54 +0000

strace (4.5.20-2) unstable; urgency=low

  * Add missing sparc syscall defines to fix FTBFS on sparc.

 -- Frederik Schüler <fs@debian.org>  Tue, 04 May 2010 14:31:46 +0200

strace (4.5.20-1) unstable; urgency=low

  [ Dmitry V. Levin ]
  * New upstream version.
    + Corrected decoding of 64bit syscalls, closes: #570603
    + Corrected getsockopt decoding on architectures where
      sizeof(long) > sizeof(int), closes: #494844

count.c
/*
 * Copyright (c) 1991, 1992 Paul Kranenburg <pk@cs.few.eur.nl>
 * Copyright (c) 1993 Branko Lankester <branko@hacktic.nl>
 * Copyright (c) 1993, 1994, 1995, 1996 Rick Sladkey <jrs@world.std.com>
 * Copyright (c) 1996-1999 Wichert Akkerman <wichert@cistron.nl>
 * Copyright (c) 1999 IBM Deutschland Entwicklung GmbH, IBM Corporation
 *                     Linux for s390 port by D.J. Barrow
 *                    <barrow_dj@mail.yahoo.com,djbarrow@de.ibm.com>
 * Copyright (c) 2004 Roland McGrath <roland@redhat.com>
 * Copyright (c) 2006 Dmitry V. Levin <ldv@altlinux.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "defs.h"

/* Per-syscall stats structure */
struct call_counts {
	/* time may be total latency or system time */
	struct timeval time;
	int calls, errors;
};

static struct call_counts *countv[SUPPORTED_PERSONALITIES];
#define counts (countv[current_personality])

static struct timeval shortest = { 1000000, 0 };

extern unsigned int delay;
extern struct timeval delay_last_tv;
extern FILE *shared_log;

void
count_syscall(struct tcb *tcp, const struct timeval *syscall_exiting_tv)
{
	struct timeval wtv;
	struct timeval *tv = &wtv;
	struct call_counts *cc;
	unsigned long scno = tcp->scno;

	if (!SCNO_IN_RANGE(scno))
		return;

	if (!counts)
		counts = xcalloc(nsyscalls, sizeof(*counts));
	cc = &counts[scno];

	cc->calls++;
	if (tcp->u_error)
		cc->errors++;

	/* tv = wall clock time spent while in syscall */
	tv_sub(tv, syscall_exiting_tv, &tcp->etime);

	/* Spent more wall clock time than spent system time? (usually yes) */
	if (tv_cmp(tv, &tcp->dtime) > 0) {
		static struct timeval one_tick = { -1, 0 };

		if (one_tick.tv_sec == -1) {
			/* Initialize it.  */
			struct itimerval it;

			memset(&it, 0, sizeof it);
			it.it_interval.tv_usec = 1;
			setitimer(ITIMER_REAL, &it, NULL);
			getitimer(ITIMER_REAL, &it);
			one_tick = it.it_interval;
//FIXME: this hack doesn't work (tested on linux-3.6.11): one_tick = 0.000000
//tprintf(" one_tick.tv_usec:%u\n", (unsigned)one_tick.tv_usec);
		}

		if (tv_nz(&tcp->dtime))
			/* tv = system time spent, if it isn't 0 */
			tv = &tcp->dtime;
		else if (tv_cmp(tv, &one_tick) > 0) {
			/* tv = smallest "sane" time interval */
			if (tv_cmp(&shortest, &one_tick) < 0)
				tv = &shortest;
			else
				tv = &one_tick;
		}
	}
	if (tv_cmp(tv, &shortest) < 0)
		shortest = *tv;
	tv_add(&cc->time, &cc->time, count_wallclock ? &wtv : tv);

	if (cflag && delay > 0)
	{
		if (delay_last_tv.tv_sec == 0 && delay_last_tv.tv_usec == 0)
		{
			delay_last_tv = *syscall_exiting_tv;
		}
		else if (delay_last_tv.tv_sec + delay <= syscall_exiting_tv->tv_sec)
		{
			delay_last_tv = *syscall_exiting_tv;

			char str[sizeof("HH:MM:SS")];
			time_t local = syscall_exiting_tv->tv_sec;
			strftime(str, sizeof(str), "%T", localtime(&local));
			fprintf(shared_log, "------------------------------%s-----------------------------\n", str);

			call_summary(shared_log);
		}
		
	}
}

static int
time_cmp(void *a, void *b)
{
	return -tv_cmp(&counts[*((int *) a)].time,
		       &counts[*((int *) b)].time);
}

static int
syscall_cmp(void *a, void *b)
{
	const char *a_name = sysent[*((int *) a)].sys_name;
	const char *b_name = sysent[*((int *) b)].sys_name;
	return strcmp(a_name ? a_name : "", b_name ? b_name : "");
}

static int
count_cmp(void *a, void *b)
{
	int     m = counts[*((int *) a)].calls;
	int     n = counts[*((int *) b)].calls;

	return (m < n) ? 1 : (m > n) ? -1 : 0;
}

static int (*sortfun)();
static struct timeval overhead = { -1, -1 };

void
set_sortby(const char *sortby)
{
	if (strcmp(sortby, "time") == 0)
		sortfun = time_cmp;
	else if (strcmp(sortby, "calls") == 0)
		sortfun = count_cmp;
	else if (strcmp(sortby, "name") == 0)
		sortfun = syscall_cmp;
	else if (strcmp(sortby, "nothing") == 0)
		sortfun = NULL;
	else {
		error_msg_and_help("invalid sortby: '%s'", sortby);
	}
}

void set_overhead(int n)
{
	overhead.tv_sec = n / 1000000;
	overhead.tv_usec = n % 1000000;
}

static void
call_summary_pers(FILE *outf)
{
	unsigned int i;
	int     call_cum, error_cum;
	struct timeval tv_cum, dtv;
	double  float_tv_cum;
	double  percent;
	const char *dashes = "----------------";
	char    error_str[sizeof(int)*3];
	int    *sorted_count;

	fprintf(outf, "%6.6s %11.11s %11.11s %9.9s %9.9s %s\n",
		"% time", "seconds", "usecs/call",
		"calls", "errors", "syscall");
	fprintf(outf, "%6.6s %11.11s %11.11s %9.9s %9.9s %s\n",
		dashes, dashes, dashes, dashes, dashes, dashes);

	sorted_count = xcalloc(sizeof(int), nsyscalls);
	call_cum = error_cum = tv_cum.tv_sec = tv_cum.tv_usec = 0;
	if (overhead.tv_sec == -1) {
		tv_mul(&overhead, &shortest, 8);
		tv_div(&overhead, &overhead, 10);
	}
	for (i = 0; i < nsyscalls; i++) {
		sorted_count[i] = i;
		if (counts == NULL || counts[i].calls == 0)
			continue;
		tv_mul(&dtv, &overhead, counts[i].calls);
		tv_sub(&counts[i].time, &counts[i].time, &dtv);
		call_cum += counts[i].calls;
		error_cum += counts[i].errors;
		tv_add(&tv_cum, &tv_cum, &counts[i].time);
	}
	float_tv_cum = tv_float(&tv_cum);
	if (counts) {
		if (sortfun)
			qsort((void *) sorted_count, nsyscalls, sizeof(int), sortfun);
		for (i = 0; i < nsyscalls; i++) {
			double float_syscall_time;
			int idx = sorted_count[i];
			struct call_counts *cc = &counts[idx];
			if (cc->calls == 0)
				continue;
			tv_div(&dtv, &cc->time, cc->calls);
			error_str[0] = '\0';
			if (cc->errors)
				sprintf(error_str, "%u", cc->errors);
			float_syscall_time = tv_float(&cc->time);
			percent = (100.0 * float_syscall_time);
			if (percent != 0.0)
				   percent /= float_tv_cum;
			/* else: float_tv_cum can be 0.0 too and we get 0/0 = NAN */
			fprintf(outf, "%6.2f %11.6f %11lu %9u %9.9s %s\n",
				percent, float_syscall_time,
				(long) (1000000 * dtv.tv_sec + dtv.tv_usec),
				cc->calls,
				error_str, sysent[idx].sys_name);

			memset(cc, 0,  sizeof(struct call_counts));
		}
	}
	free(sorted_count);

	fprintf(outf, "%6.6s %11.11s %11.11s %9.9s %9.9s %s\n",
		dashes, dashes, dashes, dashes, dashes, dashes);
	error_str[0] = '\0';
	if (error_cum)
		sprintf(error_str, "%u", error_cum);
	fprintf(outf, "%6.6s %11.6f %11.11s %9u %9.9s %s\n",
		"100.00", float_tv_cum, "",
		call_cum, error_str, "total");
}

void
call_summary(FILE *outf)
{
	unsigned int i, old_pers = current_personality;

	for (i = 0; i < SUPPORTED_PERSONALITIES; ++i) {
		if (!countv[i])
			continue;

		if (current_personality != i)
			set_personality(i);
		if (i)
			fprintf(outf,
				"System call usage summary for %d bit mode:\n",
				current_wordsize * 8);
		call_summary_pers(outf);
	}

	if (old_pers != current_personality)
		set_personality(old_pers);
}
